### PR TITLE
C Interoperability Primer

### PR DESCRIPTION
**Pull request by ShreyasKhandekar**
This is aimed at providing a primer to demonstrate some of Chapel's interoperability functionality of calling C code from Chapel and vice-versa.

The Changes made are:

- Builds upon the primer by @lydia-duncan she started in #11858 
- Adds more examples of empty`struct`s
- Examples of `proc`s using refs using a simple factorial example
- Compiling the libraries and satisfying dependencies
- Added a test for interopWithC.chpl

### CODE REVIEW

File: test/release/examples/primers/fact.c
Comment (lydia-duncan) [inline test/release/examples/primers/fact.c:18]: This is malloc'd but I don't see where it gets free'd.  Does it leak memory?

File: test/release/examples/primers/fact.c
Comment (lydia-duncan) [inline test/release/examples/primers/fact.c:24]: Ditto here

File: test/release/examples/primers/fact.c
Comment (lydia-duncan) [inline test/release/examples/primers/fact.c:37]: And here as well

File: test/release/examples/primers/fact.h
Comment (lydia-duncan) [inline test/release/examples/primers/fact.h:10]: This is a little squashed together, it'd be nice to insert some newlines.  Maybe after:
- the `#define`
- the first function definition
- the struct definition
- and the last function definition?

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:16]: We might want to rename this (and any references to it) to match the section header below it.

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:172]: Let's end this sentence with a period :) It'd also be nice to explicitly mention that we're using the keyword with a function.

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:120]: Maybe mention here that we'll see the definitions of the unfamiliar exported functions later?

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:196]: Typo on "inlcude" :)

Comment (lydia-duncan) [reply test/release/examples/primers/interopWithC.chpl:196]: I would may say "When requiring a C file, the appropriate header file must also be required." instead

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:194]: Missing period.  There may be others that I've missed

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:203]: This is quite clever :)

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:205]: I think this should have a `.. code-block:: bash` example to go with it

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:222]: This sentence is a bit run-on.  I'd maybe split it in two at the first "which", starting the second sentence with "This module defines..."

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:231]: This transition back to talking about functions is a bit abrupt.  Maybe the bit about default arguments should go before the examples of other symbols that can be declared as `extern`?

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:246]: I could see this being a bit confusing because the array is an `int` and the size is a `c_int` - it might be good to be explicit that the *array* must be of `int64_t`

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:308]: Typo on "keyowrd"

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:311]: Typo on "carefull"

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:317]: Accidental capitalization?

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [inline test/release/examples/primers/interopWithC.chpl:332]: This might warrant a `.. warning` block.  I'm also a little concerned about it being so late in the file, though maybe it is okay since LLVM is now the default.  Basically, I'm worried a new user will try to compile the file in a setting without LLVM and as a result encounter a compilation error.

We also should add a .skipif to this test so that it doesn't get tested without LLVM available

**Review by lydia-duncan — COMMENTED** [review]
This is a lot of hard work, nicely done :) Sorry if I call out things that are my fault.  There's a lot of comments, but it's a complicated thing you're adding so don't feel discouraged!

Where did you end up running this test?  Did you do a fresh checkout on our testing machines?  What about building the docs over there, too?  I like calling that sort of thing out in the PR message :) I haven't looked at the built docs for this PR myself, but I plan on doing so after you've had a round or two on the comments.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:332]: I will put it inside a `..warning` block and move it above the code block.
I feel that since this is right next to where I talk about `extern` blocks it is appropriately placed because as you said, LLVM is the default now so the average user should not encounter this issue. 
Even if a user set the target compiler to clang they will get a nice error message that says exactly what that last line says; therefore, they would not be in the dark about what went wrong.

File: doc/Makefile
Comment (ShreyasKhandekar) [reply doc/Makefile:99]: You're right, I don't think this line is necessary anymore.

**Comment by ShreyasKhandekar** [comment]
> Where did you end up running this test? Did you do a fresh checkout on our testing machines? What about building the docs over there, too? I like calling that sort of thing out in the PR message :) I haven't looked at the built docs for this PR myself, but I plan on doing so after you've had a round or two on the comments.

I was trying to run paratests but it seems I still do not have the proper unix permissions for that on our testing machines. I am working to try and resolve that. I did run a simple `start_test` on `interopWitchC.chpl` and `cClient.test.c` and they work fine.
I'll try to run more comprehensive tests once I can, I won't merge this until then.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:172]: I used the word symbol because `extern` keyword is not limited to functions. It can also apply to variables and structs. 
Should I use something other than that?

File: test/release/examples/primers/interopWithC.chpl
Comment (stonea) [inline test/release/examples/primers/interopWithC.chpl:124]: At this point I think the literal include is this:

```
foo();
alsoCallsBaz();
callUseMyType();
```

You introduced foo() above but I don't think we've introduced `alsoCallsBaz();` or `callUseMyType();` yet. Could we reorder things to introduce these earlier or just say that you'll talk about them later.

File: doc/Makefile
Comment (ShreyasKhandekar) [reply doc/Makefile:99]: ✅

File: test/release/examples/primers/fact.c
Comment (ShreyasKhandekar) [reply test/release/examples/primers/fact.c:6]: Good catch! I am trying to remove all Apache Arrow/Parquet references to avoid confusing anyone who is not familiar. I just had them in my code when I was trying a few things out.

File: test/release/examples/primers/interopWithC.chpl
Comment (stonea) [inline test/release/examples/primers/interopWithC.chpl]: Would you also need to include foo.h and the headers for whatever parts of the apache-arrow library you want to use?

File: test/release/examples/primers/interopWithC.chpl
Comment (stonea) [inline test/release/examples/primers/interopWithC.chpl]: Huh interesting. I guess I would also want to know if you can type-differentiated overloads of extern functions.

File: test/release/examples/primers/interopWithC.chpl
Comment (stonea) [inline test/release/examples/primers/interopWithC.chpl:250]: Since it looks like you can pass Chapel arrays to C I guess I'd wonder what the purpose of c_array is.

File: test/release/examples/primers/interopWithC.chpl
Comment (stonea) [inline test/release/examples/primers/interopWithC.chpl:345]: I'dd put a space between #include and <math.h>.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:332]: BTW, I also put the note about not being able to restart the Chapel runtime once it is shut down in a `.. note` block.

File: test/release/examples/primers/interopWithC.chpl
Comment (stonea) [reply test/release/examples/primers/interopWithC.chpl]: I agree with Lydia here: please add an example invocation.  When I was playing with the interop module, I guessed there were two things I needed to figure out: (1) how to write the Chapel code to declare and call the C functions and (2) how to invoke the compiler. I think your primer makes it pretty easy to figure out (1) and I don't think it's that hard to figure out (2), but someone impatiently skimming the primer to figure (2) out might just ctrl+f for "chpl" to see an example of the command.

**Review by stonea — COMMENTED** [review]
Thanks for adding this primer. I think anyone wanting to interoperate Chapel with C (or vice versa) will find this useful.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:124]: Yes, Lydia mentioned this and I've added a little statement stating that we will see how the other two functions were exported as well as their definitions later.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl]: Yes, `foo.h` must be required as well. I will modify the statement above this stating that the appropriate header file must be required when requiring a C file to say that this also applies to object files and archived libraries.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:250]: I am not sure of the differentiation myself. Maybe `c_array` was introduced before we were able to pass Chapel Arrays to C?
I can't think of why someone would still use `c_array` since Chapel Arrays are so much more powerful, unless there is a significant difference between them.
Maybe @lydia-duncan can shed some light on this.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:345]: Putting those spaces in `fact.c` as well.

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl]: I don't think that is possible currently since even when we provide default arguments, the C function will always receive all arguments when Chapel calls it.

Overloading based on types means handling arguments differently and since C does not support that I don't think we can declare such functions in Chapel either.

File: test/release/examples/primers/interopWithC.chpl
Comment (lydia-duncan) [reply test/release/examples/primers/interopWithC.chpl]: You do cover that later in the doc, but it's okay to leave it in

**Review by lydia-duncan — COMMENTED** [review]
One last typo and a couple of missing newlines.  I also don't see the free'ing of the malloc'ed memory yet?

File: test/release/examples/primers/interopWithC.chpl
Comment (ShreyasKhandekar) [reply test/release/examples/primers/interopWithC.chpl:342]: Good catch.

**Review by lydia-duncan — APPROVED** [review]
Looks good, thanks!

**Review by stonea — APPROVED** [review]

**Comment by bradcray** [comment]
Hi @ShreyasKhandekar — [caveat for the following:  I haven't read the primer nor the review comments above, so apologize if this has already been hashed through a bit or misses something].

Sorry I didn't catch this prior to the merge, but from what I've seen, I'd like to propose a modest renaming to this primer and its category.  Specifically, when I see a category with one thing in it, the first thing I think is "What else should/could be in this category ultimately?" where obvious answers for this one might be "interoperability between Chapel and Python" and/or "interoperability between Chapel and Fortran", all of which fall under the general umbrella of Interoperability.  And I think interoperability in both directions (Chapel as a library; Chapel as an executable calling standalone code) are both interesting, where the latter doesn't obviously fit into the current category.

Also, if I read the [top part of the primer](https://chapel-lang.org/docs/main/primers/interopWithC.html), it seems as though what it's teaching is how to make a library from Chapel and call it from C (while also teaching how to call C from Chapel).  So the category name seems almost like a more precise description for this primer than the primer's title itself?

To that end, I'd make the general category "Interoperating with Chapel" or "Chapel Interoperability" or possibly just "Interoperability" ("Interoperation"?).  And then I'd make the title of this primer "Calling a Chapel library from C" or "Creating Chapel libraries for C programmers" or possibly simply "Chapel<->C interoperability" (where the latter has the benefit of being of a more similar length to other primer names).

[Ultimately, I think I'd also propose that we split a "Calling from Chapel to C" primer out distinct from a "Creating a Chapel library for use by C" primer because I think the former is simpler / less prone to issues in tooling, linkers, etc. / more likely to be used in practice, and a logical stepping stone to the second primer.  But I'm not proposing you do that here, just noting it for reference].

**Comment by ShreyasKhandekar** [comment]
I think that those are more appropriate names. I will change the category title to "Interoperability" and the primer's title to "Chapel<->C Interoperability" in another pull request.

**Pull request merged**
